### ISSUE TITLE
Increase the concurrency of the S3 state backend

### ISSUE THREAD

**Issue by dmattia** [pulumi/pulumi#10057]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Some self-managed backends, such as the S3 backend, are quite slow in comparison to other backends, like the Pulumi Cloud backend. There are comments on larger issues (https://github.com/pulumi/pulumi/issues/4605#issuecomment-904477833) that reference the slowness of the S3 backend in particular, as well as a comment from a maintainer in https://github.com/pulumi/pulumi/issues/6074 that "the S3 backend is not as optimized currently as the Pulumi Service backend".

Making this more specific for this issue, I believe that one of the root causes of the slowness is that the operations when using an S3/azure state backend are written to the backend in sequence, rather than in parallel like seems to happen for the Pulumi backend.

From looking at the blob library used for talking to many of the backends, it looks like the library exposes MaxConcurrency field: https://github.com/google/go-cloud/blob/60d79299da9fb31d6bd022ae834b6766f45b55b1/blob/driver/driver.go#L65 that could be passed through in the *blob.WriterOptions of the WriteAll method: https://github.com/pulumi/pulumi/blob/ca3b9a57276751c71a4ca7fd10710edcdd72cfd3/pkg/backend/filestate/bucket.go#L22 that is used to write state checkpoint files.

Would you be open to a change where pulumi up could have a new flag named --max-state-concurrency where if set, that value would be passed through to the blob for whatever state is used? Looking through the code, it looks like Azure and S3 are the only blob backends that support it right now.

My understanding of the logic here gets pretty fuzzy though, as I can't seem to find out if that change alone would improve anything, or if there are other differences between the state backends that need to be addressed first.

Link to original Slack thread I created: https://pulumi-community.slack.com/archives/C84L4E3N1/p1657074321389129

### Affected area/feature

This would affect the S3 and Azure state backends, specifically when they call this method: https://github.com/pulumi/pulumi/blob/ca3b9a57276751c71a4ca7fd10710edcdd72cfd3/pkg/backend/filestate/bucket.go#L22

**Comment by RobbieMcKinstry** [comment]
Huh, very interesting, thank you for the write-up! I think it's unlikely we'd accept a flag for `--max-state-concurrency`. My thought is that the flag wouldn't be common enough to warrant a slot as a CLI argument.
However, I think it's _more likely_ that we'd take a PR that detects when we're writing a large volume of data to the S3 backend and select a value for `MaxConcurrency` automatically. Unfortunately, I can't promise anything there either. This might require more design. Perhaps a good starting point would be creating a benchmark, and manually toggling `MaxConcurrency` to see if you're hypothesis is correct.

**Comment by Frassle** [comment]
I'm pretty sure our writes to S3 are serial so increasing concurrency wouldn't even help. Each stack has one state file, that gets written repeatedly to the same place as a full overwrite.

**Comment by dmattia** [comment]
> We're currently on an old version of go-cloud that doesn't even support this, so first thing first would just be to update go-cloud and see how much that helps.

That sounds very promising. Is this something I could help benchmark?

**Comment by RobbieMcKinstry** [comment]
@dmattia Yes definitely! If you're able, I'd start by adding a benchmark for S3 performance. Then, I'd bump the version of gocloud, and run the benchmark again to demonstrate the change in performance. Happy to discuss alternatives if you have any other ideas. :)

**Comment by dmattia** [comment]
@RobbieMcKinstry I have some initial benchmarks as the code is now: https://github.com/dmattia/pulumi-state-backend-benchmarks/actions/runs/2639468558

For my test, I made a very basic project (see full repo: https://github.com/dmattia/pulumi-state-backend-benchmarks) that creates a Dynamodb table and then adds 500 rows to that table (each table item is a resource).

<img width="665" alt="Screen Shot 2022-07-09 at 9 30 59 AM" src="https://user-images.githubusercontent.com/8922077/178110174-359bb2bf-ed17-4adc-90f0-8d74b7d7c829.png">

Right now, it takes over 3.5x as long to provision the stack using an S3 backend compared to the Pulumi Cloud backend. Destroying the stack also takes over 2x as long.

I'll see if I can get a fork working soon to test out the newer go-cloud versions

**Comment by dmattia** [comment]
I made some changes in https://github.com/pulumi/pulumi/pull/10083 that bump up the version and set the `MaxConcurrency` field to `30` by default, and would like to use this fork to test in that repo I have above. How would I go about doing that?

**Comment by Frassle** [comment]
> How would I go about doing that?

Tricky. Your using the setup-pulumi action to install the cli and I'm pretty sure that only support installed released versions. It would be a nice suggestion to enable that to install pulumi from branches/alpha releases or something, but until that's supported I think you'd have to build a copy of the pulumi cli yourself. Either as part of the action (checkout pulumi/pulumi and run "make build" to just build the cli) or locally and pushing it somewhere to download into the action to use.

**Comment by RobbieMcKinstry** [comment]
Super cool! 

Thanks for chiming in Frassle, good advice!

**Comment by Frassle** [comment]
> So I do think it's worth getting onto the newer go-cloud versions, though I'd recommend waiting for the next go-cloud release

That sounds good to me. I know there was a PR a while a go to try and bump the version that hit some issues but I think they were issues with the specific go-cloud point release. Once the next go-cloud release comes out we can try and update to that.

**Comment by dmattia** [comment]
It might also be worth thinking about changing the default s3 backend to use `awssdk=v2`, as opposed to go-cloud defaulting to v1. I don't think there are many situations where the v2 client wouldn't be an improvement (from making the transition on a few other open source projects), but that's up to how comfortable you would be with that change

**Comment by RobbieMcKinstry** [comment]
We should look at revisiting this now that GoCloud has cut a new release.

**Comment by RobbieMcKinstry** [comment]
@dmattia just wanted to let you know that @kpitzen and I are adapting your pull request to resolve this issue. I estimate it to ship in v3.48.0. 

Gocloud defers to the AWS library to select a concurrency value when MAX_CONCURRENCY is unset. Turns out that number is `5`. We're worried that if we set the value to a fixed number, we might regress performance for some machines. For example, in your PR you set the value to 30. That might negatively adversely our Raspberry Pi users 😉. 

We're instead trying to base the value of MAX_CONCURRENCY as a multiple of [GOMAXPROCS](https://pkg.go.dev/runtime#GOMAXPROCS). That way, users with tiny machines aren't stretched too thinly, and users with beefy machines could potentially see more parallelism than _e.g._ 30.

I'm going to run some experiments to see what multipliers perform well on different resource counts. Of course, I run the risk of accidentally optimizing performance for my machine only (if I'm the one running the experiments), but maybe we can find a way to validate that performance doesn't grind to a halt when users run Pulumi on single core machines before the release.

**Comment by RobbieMcKinstry** [comment]
Just to give an update, I'm on the hook for this. I just want to run some experiments as mentioned above so we can pick a number.

**Comment by RobbieMcKinstry** [comment]
Just to give an brief update here... the delay is entirely on me! I still need to run an experiment to evaluate the optimal parameters. I've recently refreshed my benchmarking setup, so I'm hoping to pick this back up soon.

**Comment by thanh-vo-opun-ai** [comment]
Hi,
Any update for this issue?

**Comment by RobbieMcKinstry** [comment]
Hi thanh,

I'm no longer an OSS maintainer with Pulumi, so I'm afraid I can't provide any updates or advance this issue any further. @justinvp, as Core team lead, has assumed ownership of the tickets I was working on.

FWIW, before I left, I ran some experiments to determine the optimal value of this field, and I was unable to reproduce the performance wins @dmattia observed. If I had to guess, I think the issue was the machine size; I bet my experiments used a single-core machine, which may have been an issue. Admittedly, I'm not sure how core count would influence the performance of the S3 backend which is largely network-bound.

**Comment by DominicRoyStang** [comment]
@justinvp is this issue still being considered/worked on?
